python/lsst/daf/butler/registry/interfaces/_database.py

```diff
@@ -1430,10 +1432,68 @@ def delete(self, table: sqlalchemy.schema.Table, columns: Iterable[str], *rows:
             # while reporting that no rows were affected.
             return 0
         sql = table.delete()
-        whereTerms = [table.columns[name] == sqlalchemy.sql.bindparam(name) for name in columns]
-        if whereTerms:
-            sql = sql.where(sqlalchemy.sql.and_(*whereTerms))
-        return self._connection.execute(sql, *rows).rowcount
+        columns = list(columns)  # Force iterators to list
+
+        # More efficient to use IN operator if there is only one
+        # variable changing across all rows.
+        content: Dict[str, Set] = defaultdict(set)
+        if len(columns) == 1:
+            # Nothing to calculate since we can always use IN
+            column = columns[0]
+            changing_columns = [column]
+            content[column] = set([row[column] for row in rows])
+        else:
+            for row in rows:
+                for k, v in row.items():
+                    content[k].add(v)
+            changing_columns = [col for col in content if len(content[col]) > 1]
```

Comment:
` [col for col, values in content.items() if len(values) > 1]` for extra efficiency?